TYPO3.Neos/Classes/TYPO3/Neos/TypoScript/DimensionsMenuImplementation.php

```diff
@@ -0,0 +1,216 @@
+    }
+
+    /**
+     * @return string
+     */
+    public function getLabelExpression()
+    {
+        return $this->tsValue('labelExpression');
+    }
+
+    /**
+     * Builds the array of Menu items for this variant menu
+     */
+    protected function buildItems()
+    {
+        $menuItems = [];
+        $targetDimensionsToMatch = [];
+        $allDimensionPresets = $this->configurationContentDimensionPresetSource->getAllPresets();
+        $includeAllPresets = $this->getIncludeAllPresets();
+        $pinnedDimensionValues = $this->getPresets();
+
+        $pinnedDimensionName = $this->getDimension();
+        if ($pinnedDimensionName !== null) {
+            $targetDimensionsToMatch = $this->currentNode->getContext()->getTargetDimensions();
+            unset($targetDimensionsToMatch[$pinnedDimensionName]);
+        }
+
+        foreach ($this->contentDimensionCombinator->getAllAllowedCombinations() as $allowedCombination) {
+            $targetDimensions = $this->calculateTargetDimensionsForCombination($allowedCombination);
+
+            if ($pinnedDimensionName !== null && is_array($pinnedDimensionValues)) {
+                if (!in_array($targetDimensions[$pinnedDimensionName], $pinnedDimensionValues)) {
+                    continue;
+                }
+            }
+
+            // skip variants not matching the current target dimensions (except the dimension this menu covers)
+            if ($targetDimensionsToMatch !== []) {
+                foreach ($targetDimensionsToMatch as $dimensionName => $dimensionValue) {
+                    if ($targetDimensions[$dimensionName] !== $dimensionValue) {
+                        continue 2;
+                    }
+                }
+            }
+
+            $nodeInDimensions = $this->getNodeInDimensions($allowedCombination, $targetDimensions);
+
+            // no match, so we look further...
+            if ($nodeInDimensions === null && $includeAllPresets) {
+                foreach ($allowedCombination[$pinnedDimensionName] as $allowedPresetIdentifier) {
+                    $acceptableCombination = [$pinnedDimensionName => $allDimensionPresets[$pinnedDimensionName]['presets'][$allowedPresetIdentifier]['values']];
+                    $allowedAdditionalPresets = $this->configurationContentDimensionPresetSource->getAllowedDimensionPresetsAccordingToPreselection('country', [$pinnedDimensionName => $allowedPresetIdentifier]);
+                    foreach ($allowedAdditionalPresets as $allowedAdditionalDimensionName => $allowedAdditionalPreset) {
+                        $acceptableCombination[$allowedAdditionalDimensionName] = $allowedAdditionalPreset['presets'][$allowedAdditionalPreset['defaultPreset']]['values'];
+                    }
+                    $nodeInDimensions = $this->getNodeInDimensions($acceptableCombination, []);
+                    if ($nodeInDimensions !== null) {
+                        continue;
+                    }
+                }
+            }
```

Comment:
Yes. I was about to refactor things, but decided to do it as soon as the guts of this is "done".